### PR TITLE
refactor: cleanup compiler runtimes

### PR DESCRIPTION
**Pull request by bartlomieju**
* Cleanup "tsCompilerOnMessage" by factoring out separate methods for each request type:
   - "compile"
   - "runtimeCompile"
   - "runtimeTranspile"

* Simplify control flow of compiler workers by a) no longer calling "close()" in worker runtime after a 
  single message; b) explicitly shutting down worker from host after a single message

### CODE REVIEW

**Comment by bartlomieju** [comment]
CC @kitsonk

File: cli/js/compiler.ts [hunk 1]
```diff
@@ -81,259 +81,282 @@ interface CompileResult {
...
-        const configResult = host.configure(configPath, config);
-        diagnostics = processConfigureResponse(configResult, configPath);
+type RuntimeCompileResult = [
+  undefined | DiagnosticItem[],
+  Record<string, string>
+];
+
+type RuntimeBundleResult = [undefined | DiagnosticItem[], string];
+
+async function requestCompile(
```
Comment (ry) [inline cli/js/compiler.ts]: seems to me this function would be better called "compile" rather than "requestCompile"

File: cli/js/compiler.ts [hunk 2]
```diff
@@ -81,259 +81,282 @@ interface CompileResult {
...
+
+  return result;
+}
+
+async function requestRuntimeCompile(
+  request: CompilerRequestRuntimeCompile
+): Promise<RuntimeCompileResult | RuntimeBundleResult> {
+  // `RuntimeCompile` are requests from a runtime user, both compiles and
+  // bundles.  The process is similar to a request from the privileged
+  // side, but also returns the output to the on message.
```
Comment (ry) [inline cli/js/compiler.ts]: This comment would make a good jsdoc.

I feel like the function would be better called `runtimeCompile` or `compileRuntime`

Comment (kitsonk) [reply cli/js/compiler.ts]: `compileRuntime` would mean compiling the runtime, where `runtimeCompile` sounds right to me.  I agree with dropping the `request`, as these are imperatives, not sort of "pretty please do something" requests.

Comment (kitsonk) [reply cli/js/compiler.ts]: Also +1 on JSDoc! :-)

**Pull request merged**